src/ai/backend/manager/idle.py

```diff
@@ -307,24 +305,148 @@ class UtilizationIdleChecker(BaseIdleChecker):
     """
 
     name: ClassVar[str] = "utilization"
-    default_cpu_util_threshold: ClassVar[float] = 30.0
-    default_accelerator_util_threshold: ClassVar[float] = 10.0
-    default_initial_grace_period: ClassVar[float] = 300.0  # allow first 5 minutes to be idle
+    cpu_util_series: List[float] = []
+    mem_util_series: List[float] = []
+    cuda_mem_util_series: List[float] = []
 
-    async def populate_config(self, config: Mapping[str, Any]) -> None:
-        pass
+    _evhandlers: List[EventHandler[None, AbstractEvent]]
 
-    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
-        # last_stat = session['last_stat']
-        # TODO: implement
-        return True
+    async def populate_config(self, raw_config: Mapping[str, Any]) -> None:
+        self.cpu_threshold = int(raw_config["resource-thresholds"]["cpu"]["average"])
+        self.mem_threshold = int(raw_config["resource-thresholds"]["mem"]["average"])
+        self.cuda_mem_threshold = int(raw_config["resource-thresholds"]["cuda_mem"]["average"])
```

Comment:
There is "OR" check condition, but Manager fails to be launched when not all of `cpu`, `mem`, and `cuda_mem` are configured. For example, I would like to garbage collect compute sessions by CPU utilization criterion only, but I was unable to launch Manager with the following `config/idle` setting:

```json
{
    "app-streaming-packet-timeout": "5m",
    "checkers": {
        "timeout": {
            "threshold": "24h"
        },
        "utilization": {
            "resource-thresholds": {
                "cpu": {
                    "average": "10",
                    "window": "30s"
                }
            },
            "thresholds_check_condition": {
                "condition": "and"
            }
        }
    },
    "enabled": "timeout,utilization"
}
```

Reply:
Yes, I can solve this issue. I will write a lot of if/else codes to cover all kinds of configurations, unless I find a better alternative for multiple if/else code statements.

Reply:
I think you can take out the utilization calculation logic to a new internal method. Then, you may still write simpler code even if there are many independent resources to be considered. For a no-brainer mock example:

```python
cpu_util_pct = float(live_stat["cpu_util"]["pct"])
mem_util_pct = float(live_stat["mem"]["pct"])
cuda_mem_util_pct = float(live_stat["cuda_mem"]["pct"])

cpu_check_result = _check_resource('cpu', cpu_util_pct)
mem_check_result = _check_resource('mem', mem_util_pct)
cuda_mem_check_result = _check_resource('cuda_mem', cuda_mem_util_pct)
# If async function is needed, you may create tasks and use asyncio.gather to collect every results

if self.threshold_condition == "or":
  if cpu_check_result or mem_check_result or cuda_mem_check_result:
    return ...
...

def _check_resource(util_series_name, util_pct):
  interval = self.timer.interval
  window_size = ...
  util_series = getatr(self, f'{util_series_name}')
  if len(util_series) < window_size:
    setattr(self, f'{util_series_name}_util_series', util_series.append(float(cpu_util_pct)))
    return True
  avg_util = ...
```

Reply:
Also, for easier fetching of values from deeply-nested dict, you can use `nmget` function in `common` module.
https://github.com/lablup/backend.ai-common/blob/main/src/ai/backend/common/utils.py#L105

```python
from ai.backend.common.utils import nmget
self.cpu_threshold = int(nmget(raw_config, "resource-thresholds.cpu.average"))
self.mem_threshold = int(nmget(raw_config, "resource-thresholds.mem.average"))
self.cuda_mem_threshold = int(nmget(raw_config, "resource-thresholds.cuda_mem.average"))
```